File: src/include/fetch/WebItemReference.php
<?php

namespace fetch;

use DateTime;
use Exception;
use GuzzleHttp\Client;
use GuzzleHttp\Exception\ClientException;
use log\Log;
use log\LogLevel;

class WebItemReference
{

	/**
	 * @var array<string, string>
	 */
	private static array $defaultHeaders = [
		"User-Agent" => "3dassets.one / Fetching"
	];


	/**
	 * 
	 * @var array<string,mixed>
	 */
	private array $options = [];

	/**
	 * 
	 * @param string $url 
	 * @param string $method 
	 * @param array<string,string> $headers 
	 * @param string $requestBody 
	 * @param array<string,string> $queryParameters 
	 * @return void 
	 */
	public function __construct(
		public string $url,
		public string $method = 'GET',
		public array $headers = [],
		public string $requestBody = "",
		public array $queryParameters = []
	) {
		$this->headers = array_merge(self::$defaultHeaders, $this->headers);

		$this->options = [
			'headers' => $this->headers,
		];

		// Add body or query parameters if provided
		if (!empty($this->requestBody)) {
			$this->options['body'] = $this->requestBody;
		} elseif (!empty($this->queryParameters)) {
			$this->options['form_params'] = $this->queryParameters;
		}
	}

	public function fetchCookie(string $targetCookieName): ?string
	{
		Log::write("Fetching cookie for request: ", ["targetCookieName" => $targetCookieName, "request" => $this], LogLevel::DEBUG);

		$client = new Client(['cookies' => true]);
		try {
			$client->request($this->method, $this->url, $this->options);

			// https://github.com/guzzle/guzzle/issues/3114#issuecomment-1627228395
			$cookieJar = $client->getConfig('cookies');
			$cookie = $cookieJar->getCookieByName($targetCookieName)->getValue();
			Log::write("Cookie Request successful!", LogLevel::DEBUG);
		} catch (ClientException $e) {
			Log::write("Cookie Request client error", [$e->getCode(), $e->getMessage()], LogLevel::ERROR);
			$cookie = NULL;
		} catch (Exception $e) {
			Log::write("Cookie request generic error: ", [$e->getCode(), $e->getMessage()], LogLevel::ERROR);
			$cookie = NULL;
		}

		Log::write("Cookie value determined", ["cookie" => $cookie], LogLevel::DEBUG);

		return $cookie; // Return null if the target cookie was not found
	}

	public function fetch(): FetchedWebItem
	{
		$client = new Client();
		Log::write("Fetching: ", ["request" => $this], LogLevel::DEBUG);
		try {
			$result = $client->request($this->method, $this->url, $this->options);
			$content = $result->getBody();
		} catch (ClientException $e) {
			Log::write("Request client error", [$e->getCode(), $e->getMessage()], LogLevel::ERROR);
			$content = NULL;
		} catch (Exception $e) {
			Log::write("Generic request error: ", [$e->getCode(), $e->getMessage()], LogLevel::ERROR);
			$content = NULL;
		}

		Log::write("Request completed", ["length" => strlen($content ?? ""), "statusCode" => isset($result) ? $result->getStatusCode() : NULL], LogLevel::INFO);

		return new FetchedWebItem(
			reference: $this,
			content: $content,
			httpStatusCode: isset($result) ? $result->getStatusCode() : NULL,
			lastUpdated: NULL
		);
	}
}

File: app/fetch/web_item_reference.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Optional

import httpx

from app.log.log import Log
from app.log.log_level import LogLevel


@dataclass
class WebItemReference:
    url: str
    method: str = "GET"
    headers: dict[str, str] = field(default_factory=dict)
    request_body: str = ""
    query_parameters: dict[str, str] = field(default_factory=dict)

    DEFAULT_HEADERS: dict[str, str] = field(
        default_factory=lambda: {"User-Agent": "3dassets.one / Fetching"},
        init=False,
        repr=False,
    )

    def __post_init__(self) -> None:
        default_hdrs = {"User-Agent": "3dassets.one / Fetching"}
        self.headers = {**default_hdrs, **self.headers}

    def fetch(self) -> "FetchedWebItem":
        from app.fetch.fetched_web_item import FetchedWebItem

        Log.write("Fetching", {"url": self.url, "method": self.method}, LogLevel.INFO)
        try:
            kwargs: dict = {
                "headers": self.headers,
                "follow_redirects": True,
                "timeout": 30.0,
            }
            if self.request_body:
                kwargs["content"] = self.request_body.encode()
            elif self.query_parameters:
                kwargs["data"] = self.query_parameters

            with httpx.Client() as client:
                response = client.request(self.method, self.url, **kwargs)
            content: Optional[bytes] = response.content
            status_code: Optional[int] = response.status_code
            Log.write("Request completed", {"status": status_code, "length": len(content or b"")}, LogLevel.INFO)
        except httpx.HTTPStatusError as e:
            Log.write("HTTP error", {"code": e.response.status_code, "message": str(e)}, LogLevel.ERROR)
            content = None
            status_code = e.response.status_code
        except Exception as e:
            Log.write("Request error", str(e), LogLevel.ERROR)
            content = None
            status_code = None

        return FetchedWebItem(reference=self, content=content, http_status_code=status_code)

    def fetch_cookie(self, target_cookie_name: str) -> Optional[str]:
        Log.write("Fetching cookie", {"target": target_cookie_name, "url": self.url}, LogLevel.INFO)
        try:
            kwargs: dict = {
                "headers": self.headers,
                "follow_redirects": True,
                "timeout": 30.0,
            }
            with httpx.Client() as client:
                client.request(self.method, self.url, **kwargs)
                cookie_value = client.cookies.get(target_cookie_name)
            Log.write("Cookie fetched", {"cookie": cookie_value}, LogLevel.DEBUG)
            return cookie_value
        except Exception as e:
            Log.write("Cookie request error", str(e), LogLevel.ERROR)
            return None
